pkg/kubelet/apis/pluginregistration/v1alpha/api.proto

```diff
@@ -0,0 +1,45 @@
+// To regenerate api.pb.go run hack/update-generated-kubelet-plugin-registration.sh
+syntax = 'proto3';
+
+package pluginregistration;
+
+import "github.com/gogo/protobuf/gogoproto/gogo.proto";
+
+option (gogoproto.goproto_stringer_all) = false;
+option (gogoproto.stringer_all) =  true;
+option (gogoproto.goproto_getters_all) = true;
+option (gogoproto.marshaler_all) = true;
+option (gogoproto.sizer_all) = true;
+option (gogoproto.unmarshaler_all) = true;
+option (gogoproto.goproto_unrecognized_all) = false;
+
+// PluginInfo is the message sent from a plugin to the Kubelet pluginwatcher for plugin registration
+message PluginInfo {
+	// Type of the Plugin. CSIPlugin or DevicePlugin
+	string type = 1;
+	// Plugin name that uniquely identifies the plugin for the given plugin type.
+	// For DevicePlugin, this is the resource name that the plugin manages and
+	// should follow the extended resource name convention.
+	// For CSI, this is the CSI driver name.
+	string name = 2;
+	// Plugin service API versions the plugin supports.
+	// For DevicePlugin, this maps to the deviceplugin API versions the
+	// plugin supports at the given socket.
+	// The Kubelet component communicating with the plugin should be able
+	// to choose any preferred version from this list, or returns an error
+	// if none of the listed versions is supported.
+	repeated string version = 3;
+}
+
+// RegistrationStatus is the message sent from Kubelet pluginwatcher to the plugin for notification on registration status
+message RegistrationStatus {
+	// True if plugin gets registered successfully at Kubelet
+	bool plugin_registered  = 1;
+	// Error message in case plugin fails to register, empty string otherwise
+	string error  = 2;
+}
+
+// Registration is the service advertised by the Plugins.
+service Registration {
+	rpc GetInfo(stream RegistrationStatus) returns (stream PluginInfo) {}
```

Comment:
This stream handshake still seem weird
And seem to introduce some code complexity with regards to handling streams correctly and closing the stream.

I think it would make more sens to use the previous iteration with a simple:
```
service GetInfo(PluginIdentityRequest) returns (PluginIdentityResponse) {}
service PluginRegistrationStatus(PluginRegistrationStatus) returns (Empty) {}
```

Reply:
I would prefer bi-directional stream because it requires single RPC call only. Complexity depends on perception and is subjective. It varies from person to person.

Reply:
> I would prefer bi-directional stream because it requires single RPC call only.

And what does that single RPC implementation bring that's simpler?
Because I see a clear value in design (separation of concerns) and code size (less code) with the double RPC implementation

Reply:
I see single bidirectional RPC logical because I visualize it as a single device registration operation. Plugin sends info and then waits for registration response back. I dont understand what you meant by seperation of concerns. I see plugin's concern is to know whether it got registered or not. And for registration only, plugin did send info. Where are separate concerns?

Reply:
>This stream handshake still seem weird

This statement seems wierd.

Reply:
A close analogy to bi-directional scheme here would be TCP handshake. There are pros and cons for both approaches.
Whichever one we choose let's be consistent not just here, but in the plugin APIs as well.
cc @saad-ali @jiayingz

Reply:
I actually continued this in a later comment here are my thoughts:

Regarding the bi-directional stream, I don't understand what a stream brings when we are sending a single message.
The added complexity in the Kubelet implementation in terms of:
- added code (stream handling, added code, handling of unexpected client messages)
- architecture, there are two concerns here: 
    - The identity of the plugin
    - the registration status 

Additionnally the API clearly doesn't really make it obvious why this is a stream (when it's a stream I expect multiple messages) and how many I should get.
```protobuf
service Registration {
	rpc GetInfo(stream RegistrationStatus) returns (stream PluginInfo) {}
}

service Registration {
     service GetInfo(PluginInfoRequest) returns (PluginInfoResponse) {}
     service PluginRegistrationStatus(PluginRegistrationStatus) returns (Empty) {}
}
```

Finally is pretty obvious when you compare both implementation that one is simpler and more explicit than the other: 
**Device Plugin Stream**
```golang
func (e *examplePlugin) GetInfo(stream watcherapi.Registration_GetInfoServer) error {
	info := watcherapi.PluginInfo{
		Type:              "example-plugin-type",
		Name:              "example-plugin",
		SupportedVersions: []string{"v1beta1", "v1beta2"},
	}
	err := stream.Send(&info)
	if err != nil {
		glog.Errorf("Error sending plugin info: %v\n", err)
		return err
	}

	s, err := stream.Recv()
	if err != nil {
		glog.Errorf("Received err: %v\n", err)
		return err
	}
	glog.Infof("Got stream")
	fmt.Println("stream:", s)
	if e.registrationStatus != nil {
		e.registrationStatus <- *s
	}
	if s.PluginRegistered {
		return nil
	} else {
		glog.Errorf("Registration failed: %s\n", s.Error)
		return fmt.Errorf("Registration failed: %s", s.Error)
	}
}

```

**Device Plugin Double RPC**
```golang
func (e *examplePlugin) GetInfo(req *watcherapi.PluginInfoRequest) (*watcherapi.PluginInfoResponse, error) {
	return &watcherapi.PluginInfoResponse{
		Type:              "example-plugin-type",
		Name:              "example-plugin",
		SupportedVersions: []string{"v1beta1", "v1beta2"},
	}, nil
}

func (e *examplePlugin) PluginRegistrationStatus(status *watcherapi.PluginIdentityRequest) (*watcherapi.Empty, error) {
	if e.registrationStatus != nil {
		e.registrationStatus <- *s
	}

	if !s.PluginRegistered {
		glog.Errorf("Registration failed: %s\n", s.Error)
	}

	return &watcherapi.Empty{}, nil
}
```
**Kubelet Stream**
```golang
func (w *Watcher) registerPlugin(socketPath string) error {
	//TODO: Implement rate limiting to mitigate any DOS kind of attacks.
	glog.V(2).Infof("registerPlugin called for socketPath: %s", socketPath)
	client, conn, err := dial(socketPath)
	if err != nil {
		glog.Warningf("Dial failed at socket %s, err: %v", socketPath, err)
		return err
	}
	defer conn.Close()
	stream, err := client.GetInfo(context.Background())
	if err != nil {
		glog.Errorf("Failed to start GetInfo grpc stream at socket %s, err: %v", socketPath, err)
		return err
	}
	defer stream.CloseSend()
	rqst, err := stream.Recv()
	if err != nil {
		glog.Errorf("Failed to receive stream at socket %s, err: %v", socketPath, err)
		return err
	}

	// hiding the callback handling under a function call
	resp := RegistrationStatusCallback(rqst)

	glog.V(2).Infof("send register resp: %v", resp)
	if err := stream.Send(resp); err != nil {
		glog.Errorf("Failed to send registration status at socket %s, err: %v", socketPath, err)
		return err
	}

	for {
		_, err := stream.Recv()
		if err == io.EOF {
			break
		} else if err != nil {
			glog.Errorf("Recv failed. err: %v", err)
			return err
		}
	}

	if !resp.PluginRegistered {
		glog.Errorf("Registration failed for plugin type: %s, name: %s, socket: %s, err: %s", rqst.Type, rqst.Name, socketPath, resp.Error)
		return fmt.Errorf("%s", resp.Error)
	}
	glog.Infof("Successfully registered plugin for plugin type: %s, name: %s, socket: %s", rqst.Type, rqst.Name, socketPath)
	return nil
}
```
**Kubelet Double RPC**
```golang
func (w *Watcher) registerPlugin(socketPath string) error {
	//TODO: Implement rate limiting to mitigate any DOS kind of attacks.
	glog.V(2).Infof("registerPlugin called for socketPath: %s", socketPath)
	client, conn, err := dial(socketPath)
	if err != nil {
		glog.Warningf("Dial failed at socket %s, err: %v", socketPath, err)
		return err
	}
	defer conn.Close()

	info, err := client.GetInfo(context.Background(), &watcherapi.PluginInfoRequest{})
	if err != nil {
		glog.Errorf("Failed to GetInfo for socket %s, err: %v", socketPath, err)
		return err
	}

	// hiding the callback handling under a function call
	resp := RegistrationStatusCallback(info)

	identity, err := client.PluginRegistrationStatus(context.Background(), resp)
	if err != nil {
		glog.Errorf("Failed to send PluginRegistrationStatus at socket %s, err: %v", socketPath, err)
		return err
	}

	if !resp.PluginRegistered {
		glog.Errorf("Registration failed for plugin type: %s, name: %s, socket: %s, err: %s", rqst.Type, rqst.Name, socketPath, resp.Error)
		return fmt.Errorf("%s", resp.Error)
	}

	glog.Infof("Successfully registered plugin for plugin type: %s, name: %s, socket: %s", rqst.Type, rqst.Name, socketPath)
	return nil
}
```